src/libertem/analysis/clust.py

```diff
@@ -13,9 +14,58 @@
 from libertem.udf.stddev import StdDevUDF, consolidate_result
 from libertem import masks
 from libertem.masks import _make_circular_mask
-
-
-class ClusterAnalysis(BaseAnalysis):
+from .helper import GeneratorHelper
+
+
+class ClusterTemplate(GeneratorHelper):
+
+    short_name = "cluster"
+
+    def __init__(self, params):
+        self.params = params
+
+    def get_dependency(self):
+        dep = ["from libertem.analysis import ClusterAnalysis",
+               "from libertem.udf.stddev import StdDevUDF, consolidate_result",
+               "from libertem.masks import _make_circular_mask",
+               "from skimage.feature import peak_local_max",
+               "import sparse",
+               "from libertem.udf.masks import ApplyMasksUDF"]
+        return dep
+
+    def temp_cluster_controller(self):
```

Comment:
Is this what you are suggesting @uellue , we have to care about template when changing the `controller`.

Reply:
That's already an improvement since it is more likely to break if `controller()` is changed without synchronizing with `temp_cluster_controller()` as soon as the generated notebooks are run in CI.

Even better would be to refactor `controller()`. The text replacement is already a good start.

Possible ideas:

* Move as much functionality as reasonably possible from `controller()` to individual helper methods of the analysis.
  * Instantiating UDFs
  * Interpreting parameters
  * Interpreting UDF results
* Parametrize aspects that are different, such as `self` vs. a variable for the analysis object
* Turn `controller()` into a wrapper that mostly calls a function that looks similar to the current `temp_cluster_controller()` output, i.e. with `self` as a parameter for the analysis, same for dataset etc.

The main points where the controller and the notebook script contents are inherently different is the method for running UDFs: sync vs. async. It seems that the analysis doesn't do too much interesting stuff here, apparently just running the UDF and getting the result in the end. @sk1p can that perhaps be parametrized or abstracted away somehow, ideally that there is just a `ctx.run_udf(...)` in the source code and we somehow mock what `ctx` is in this case to that it works when we call from `controller()` in the async case?

Reply:
Anyway, that's a bit fiddly and better left for the second iteration. From my gut feeling this will be solvable in the current framework of your code one way or another, which is the main point: Changes here will not affect the notebook implementation in the other analyses.